File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root == None:
            return []

        q = deque([(root, 0, 0)])

        nodes = defaultdict(lambda: defaultdict(list))

        while q:
            node, row, col = q.popleft()
            nodes[col][row].append(node.val)

            if node.left:
                q.append((node.left, row + 1, col - 1))
            if node.right:
                q.append((node.right, row + 1, col + 1))

        ans = []

        for col in sorted(nodes):
            cur = []
            for row in sorted(nodes[col]):
                cur.extend(sorted(nodes[col][row]))
            ans.append(cur)
        
        return ans

File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<int, map<int, multiset<int>>> st;
    void dfs(TreeNode* root, int row, int col) {
        if (!root) return ;

        st[col][row].insert(root -> val);
        dfs(root -> left, row + 1, col - 1);
        dfs(root -> right, row + 1, col + 1);
    }
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        vector<vector<int>> ans;

        dfs(root, 0, 0);

        for (auto &p : st) {
            vector<int> level;

            for (auto &i : p.second) {
                level.insert(level.end(), i.second.begin(), i.second.end());
            }

            ans.push_back(level);
        }

        return ans;
    }
};